rust/headless-client/src/linux.rs

```diff
@@ -205,29 +207,39 @@ async fn ipc_listen(sock_path: &Path) -> Result<()> {
             continue;
         }
 
+        let stream = IpcStream::new(stream, LengthDelimitedCodec::new());
         if let Err(error) = handle_ipc_client(stream).await {
             tracing::error!(?error, "Error while handling IPC client");
         }
     }
 }
 
-async fn handle_ipc_client(mut stream: UnixStream) -> Result<()> {
+type IpcStream = tokio_util::codec::Framed<UnixStream, LengthDelimitedCodec>;
+
+async fn handle_ipc_client(mut stream: IpcStream) -> Result<()> {
     tracing::info!("Waiting for an IPC message from the GUI...");
-    let v = read_ipc_msg(&mut stream).await?;
-    let s = String::from_utf8(v)?;
+
+    let v = stream
+        .next()
+        .await
+        .context("Error while reading IPC message")?
+        .context("IPC stream empty")?;
+    let s = String::from_utf8(v.to_vec())?;
```

Comment:
I'd recommend to make your codec that composes `LengthDelimitedCodec` and uses `serde_json` to go from bytes to messages.

Reply:
Also, there is no need for this allocation, `serde_json` can deserialize from bytes directly.

Reply:
I tried making a new struct that wraps all this and implements `Sink` and `Stream` but it seemed like too much work since we're the only user of it. We could do it later on.

switched to `serde_json::from_slice`, I didn't know it existed. I would have guessed `from_bytes`.

Reply:
> I tried making a new struct that wraps all this and implements `Sink` and `Stream` but it seemed like too much work since we're the only user of it. We could do it later on.

The trick is that you don't implement `Sink` and `Stream` but `Encoder` and `Decoder` and compose what you want, i.e. first call your length-delimited codec and once you have a chunk of bytes try to deserialize JSON from it. You then get `Sink` and `Stream` by wrapping your `AsyncRead + AsyncWrite` with a `Framed` and your custom codec, the same way you currently initialize it with just the length-delimited one.